core/variable_map.js

```diff
@@ -0,0 +1,222 @@
+ *     a UUID.
+ */
+Blockly.VariableMap.prototype.createVariable = function(name, opt_type, opt_id) {
+  var variable = this.getVariable(name);
+  if (variable) {
+    if (opt_type && variable.type != opt_type) {
+      throw Error('Variable "' + name + '" is already in use and its type is "'
+                  + variable.type + '" which conflicts with the passed in ' +
+                  'type, "' + opt_type + '".');
+    }
+    if (opt_id && variable.getId() != opt_id) {
+      throw Error('Variable "' + name + '" is already in use and its id is "'
+                  + variable.getId() + '" which conflicts with the passed in ' +
+                  'id, "' + opt_id + '".');
+    }
+    return;
+  }
+  if (opt_id && this.getVariableById(opt_id)) {
+    throw Error('Variable id, "' + opt_id + '", is already in use.');
+  }
+  opt_id = opt_id || Blockly.utils.genUid();
+  opt_type = opt_type || '';
+
+  variable = new Blockly.VariableModel(name, opt_type, opt_id);
+  // If opt_type is not a key, create a new list.
+  if (!this.variableMap_[opt_type]) {
+    this.variableMap_[opt_type] = [variable];
+  }
+  // Else append the variable to the preexisting list.
+  else {
+    this.variableMap_[opt_type].push(variable);
+  }
+};
+
+/**
+ * Delete a variable.
+ * @param {Blockly.VariableModel} variable Variable to delete.
+ */
+Blockly.VariableMap.prototype.deleteVariable = function(variable) {
+  var type = variable.type;
+  for (var i = 0, tempVar; tempVar = this.variableMap_[type][i]; i++) {
+    if (Blockly.Names.equals(tempVar.name, variable.name)) {
+      delete this.variableMap_[type][i];
+      this.variableMap_[type].splice(i, 1);
+      return;
+    }
+  }
+};
+
+/**
+ * Find the variable by the given name and return it. Return null if it is not
+ *     found.
+ * @param {!string} name The name to check for.
+ * @return {?Blockly.VariableModel} the variable with the given name.
+ */
+Blockly.VariableMap.prototype.getVariable = function(name) {
+  var keys = Object.keys(this.variableMap_);
+  for (var i = 0; i < keys.length; i++ ) {
+    var key = keys[i];
+    for (var j = 0, variable; variable = this.variableMap_[key][j]; j++) {
+      if (Blockly.Names.equals(variable.name, name)) {
+        return variable;
+      }
+    }
+  }
+  return null;
+};
+
+/**
+ * Find the variable by the given id and return it. Return null if it is not
+ *     found.
+ * @param {!string} id The id to check for.
+ * @return {?Blockly.VariableModel} The variable with the given id.
+ */
+Blockly.VariableMap.prototype.getVariableById = function(id) {
+  var keys = Object.keys(this.variableMap_);
+  for (var i = 0; i < keys.length; i++ ) {
+    var key = keys[i];
+    for (var j = 0, variable; variable = this.variableMap_[key][j]; j++) {
+      if (variable.getId() == id) {
+        return variable;
+      }
+    }
+  }
+  return null;
+};
+
+/**
+ * Find the variable with the specified type. If type is null, return list of
```

Comment:
"Get a list containing all of the variables of a specified type.  If type is null, ..."